hbase-server/src/main/java/org/apache/hadoop/hbase/wal/BaseRecoveredEditsOutputSink.java

```diff
@@ -0,0 +1,281 @@
+    // If we already decided that this region doesn't get any output
+    // we don't need to check again.
+    if (blacklistedRegions.contains(region)) {
+      return null;
+    }
+    String tmpDirName = walSplitter.conf.get(HConstants.TEMPORARY_FS_DIRECTORY_KEY,
+        HConstants.DEFAULT_TEMPORARY_HDFS_DIRECTORY);
+    Path regionedits = getRegionSplitEditsPath(entry,
+        walSplitter.getFileBeingSplit().getPath().getName(), tmpDirName, conf);
+    if (regionedits == null) {
+      blacklistedRegions.add(region);
+      return null;
+    }
+    FileSystem walFs = FSUtils.getWALFileSystem(conf);
+    if (walFs.exists(regionedits)) {
+      LOG.warn("Found old edits file. It could be the "
+          + "result of a previous failed split attempt. Deleting " + regionedits + ", length="
+          + walFs.getFileStatus(regionedits).getLen());
+      if (!walFs.delete(regionedits, false)) {
+        LOG.warn("Failed delete of old {}", regionedits);
+      }
+    }
+    WALProvider.Writer w = walSplitter.createWriter(regionedits);
+    LOG.info("Creating recovered edits writer path={}", regionedits);
+    return new RecoveredEditsWriter(regionedits, w, entry.getKey().getSequenceId());
+  }
+
+  protected Path closeWriter(String encodedRegionName, RecoveredEditsWriter wap,
+      List<IOException> thrown) throws IOException {
+    try {
+      wap.writer.close();
+    } catch (IOException ioe) {
+      LOG.error("Could not close recovered edits at {}", wap.path, ioe);
+      thrown.add(ioe);
+      return null;
+    }
+    LOG.info("Closed recovered edits writer path={} (wrote {} edits, skipped {} edits in {} ms",
+        wap.path, wap.editsWritten, wap.editsSkipped, wap.nanosSpent / 1000 / 1000);
+    if (wap.editsWritten == 0) {
+      // just remove the empty recovered.edits file
+      if (walFS.exists(wap.path) && !walFS.delete(wap.path, false)) {
+        LOG.warn("Failed deleting empty {}", wap.path);
+        throw new IOException("Failed deleting empty  " + wap.path);
+      }
+      return null;
+    }
+
+    Path dst = getCompletedRecoveredEditsFilePath(wap.path,
+        regionMaximumEditLogSeqNum.get(encodedRegionName));
+    try {
+      if (!dst.equals(wap.path) && walFS.exists(dst)) {
+        deleteOneWithFewerEntries(wap, dst);
+      }
+      // Skip the unit tests which create a splitter that reads and
+      // writes the data without touching disk.
+      // TestHLogSplit#testThreading is an example.
+      if (walFS.exists(wap.path)) {
+        if (!walFS.rename(wap.path, dst)) {
+          throw new IOException("Failed renaming recovered edits " + wap.path + " to " + dst);
+        }
+        LOG.info("Rename recovered edits {} to {}", wap.path, dst);
+      }
+    } catch (IOException ioe) {
+      LOG.error("Could not rename recovered edits {} to {}", wap.path, dst, ioe);
+      thrown.add(ioe);
+      return null;
+    }
+    return dst;
+  }
+
+  @Override
+  public boolean keepRegionEvent(WAL.Entry entry) {
```

Comment:
I need to see what this method is for. Its an odd name.

Reply:
It is a abstract method in OutputSink. Here is an impl and only keep the compaction event. It was used in WALSplitter#splitLogFile method. 
`// Don't send Compaction/Close/Open region events to recovered edit type sinks.`
`        if (entry.getEdit().isMetaEdit() && !outputSink.keepRegionEvent(entry)) {`
`          editsSkipped++;`
`          continue;`
`        }`
`       entryBuffers.appendEntry(entry);`
It need the caller to call isMetaEdit first...